### DISCUSSION TITLE
How can I install additional packages using a file preinstall.sh?

### DISCUSSION THREAD

**Discussion by AlexanderLevashov** [goreleaser/nfpm#586] [Q&A]
Hello everyone. There is a specific problem. I need to install packages via preinstall, but the problem is that nfpm cannot unpack the archive. The following errors occur:

`Reading package lists... Done
Building dependency tree       
Reading state information... Done
Note, selecting 'hello' instead of './hello_1.0.3_amd64.deb'
The following NEW packages will be installed:
  hello
0 upgraded, 1 newly installed, 0 to remove and 642 not upgraded.
After this operation, 1 212 kB of additional disk space will be used.
Get:1 /home/alexander/hello_test/hello_1.0.3_amd64.deb hello amd64 1.0.3 [570 kB]
(Reading database ... 210508 files and directories currently installed.)
Preparing to unpack .../hello_1.0.3_amd64.deb ...
tar (child): php-8.1.13.tar.gz: Cannot open: No such file or directory
tar (child): Error is not recoverable: exiting now
tar: Child returned status 2
tar: Error is not recoverable: exiting now
dpkg: error processing archive /home/alexander/hello_test/hello_1.0.3_amd64.deb 
(--unpack):
 new hello package pre-installation script subprocess returned error exit status
 2
Errors were encountered while processing:
 /home/alexander/hello_test/hello_1.0.3_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)`

I know that the missing packages can be registered in depends in nfpm.yaml, but I need it through the preinstall process
As an example in preinstall.sh I have only tar -xvzf 8.1.13.tar.gz (just an example). And at the same time, if you make this file executable, then everything works, but not through nfpm. I ask for help, thank you.

**Comment by caarlos0** [comment]
can you share a minimum reproducible of this?

**Reply by AlexanderLevashov** [reply]
In general, as I understood, I need to fix the errors: tar (child): php-8.1.13.tar.gz : Cannot open (I do not know why this does not work, although if you make preinstall executable and run it ./preinstall.sh - then everything works fine in this case. But if you run the compiled deb, it gives these errors), you also need to fix the error; Sub-process /usr/bin/dpkg returned an error code (1).
And I don't have such a config file
I'm launching
```
nfpm pkg --package deb --target ./
sudo apt install ./hello_1.0.3_amd64.deb
```

and as a result, these errors appear

**Reply by caarlos0** [reply]
I understood that you got that error, still, can you share the minimal configurations/files needed to reproduce the error?

https://stackoverflow.com/help/minimal-reproducible-example

**Reply by AlexanderLevashov** [reply]
@caarlos0, thanks for the answers, I have already solved the question.

**Reply by caarlos0** [reply]
can you share what it was? it might be helpful to other people too :)

**Reply by AlexanderLevashov** [reply]
@caarlos0, sorry for long answer. 
It's my own fault. For me, nfpm and the build of the distribution is basically a new thing, I haven't done this before. Firstly, I incorrectly specified the path in the files with the sh extension - this is the main error. 
But that's not all. After I pointed out the right ways, another error occurred, this is secondly. T
he error was as follows, during the installation of the hello deb package, when the process reached the installation of dependencies (I used docker) then the following error occurs: 
`E: Could not get lock /var/lib/dpkg/lock-frontend - open (11: Resource temporarily unavailable)`
I solved this problem by writing an additional terminal call in the bash script, through which the command would be run: in my case; 
`sudo gnome-terminal --tab --title="Install_test" --command="bash -c'sudo apt install /docker.io_20.10.12-0ubuntu2~20.04.1_amd64.deb;'"`

**Reply by AlexanderLevashov** [reply]
@caarlos0, I have another problem. Everything I described above, of course, works, dependencies are installed, but it looks very ugly, especially when an additional console is called.

The problem is this, there is an online machine that has Internet access and there is an offline machine that does not.

How do I build a deb package so that it immediately has a distribution, for example, the same docker. Because there will be no internet on the other machine.
Eventually run the command `sudo apt install ./hello.deb `
And so that both hello and docker are installed in the end. How do I do this?
